libphobos/libdruntime/core/atomic.d

```diff
@@ -1415,6 +1420,11 @@ else version( GNU )
             res = __atomic_compare_exchange_8(cast(void*) here, cast(void*) &ifThis, writeThis,
                                               false, MemoryOrder.seq, MemoryOrder.seq);
         }
+        else static if( T.sizeof == long.sizeof * 2)
+        {
+            res = __atomic_compare_exchange(T.sizeof, cast(void*)here, cast(void*)&ifThis, cast(void*)&writeThis,
+                                            MemoryOrder.seq, MemoryOrder.seq);
```

Comment:
GCC documents this function as `__atomic_exchange (type *ptr, type *val, type *ret, int memorder)` (https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html) and turning the typed version into the `_X` or the untyped generic fallback indeed requires frontend support. `__atomic_exchange (size_t length, void *ptr, void *val, type *ret, int memorder)` is the generic implementation in libatomic (https://github.com/gcc-mirror/gcc/blob/master/libatomic/libatomic_i.h#L223).

I don't think we have to add this frontend logic though. As atomic calls should go through `core.atomic` anyway, we can just make sure to call the correct `_X` builtins. In this special case we should probably make calling the `_16` versions work.

But the main question is whether we actually accept `libatomic` as a new dependency for druntime. I guess the only other solutions are not implementing 128 bit atomics at all or porting DMDs inline ASM.